pkg/ipamd/datastore/data_store.go

```diff
@@ -883,8 +1187,56 @@ func (ds *DataStore) GetENIIPs(eniID string) ([]string, error) {
 	}
 
 	var ipPool = make([]string, 0, len(eni.IPv4Addresses))
-	for ip := range eni.IPv4Addresses {
+	for ip, addr := range eni.IPv4Addresses {
+		// /32 IPs from SIP or PD pool will be in eni.IPv4Addresses
+		// hence this has to be checked, IP belonging to a prefix can be
+		// a possibility because of PD enable/disable knob toggle
+		if !eni.IsPDenabled && addr.Prefix != "" {
+			ds.log.Debugf("IP %s belongs to prefix pool so do not account", ip)
+			continue
+		}
 		ipPool = append(ipPool, ip)
 	}
 	return ipPool, nil
 }
+
+// GetENIPrefixes returns the known (allocated & unallocated) ENI Prefixed.
+func (ds *DataStore) GetENIPrefixes(eniID string) ([]string, error) {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	eni, ok := ds.eniPool[eniID]
+	if !ok {
+		return nil, errors.New(UnknownENIError)
+	}
+
+	var ipPool = make([]string, 0, len(eni.IPv4Prefixes))
+	for prefix, prefixData := range eni.IPv4Prefixes {
+		ip := (prefix + "/" + strconv.Itoa(prefixData.PrefixLen))
+		ipPool = append(ipPool, ip)
+	}
+	return ipPool, nil
+}
+
+func (ds *DataStore) GetFreePrefixes() int {
+	freePrefixes := 0
```

Comment:
I think we should hold the datastore lock here.  I agree we probably don't need it when walking the structures as we do (golang reference counting will save us from walking an object that other code frees concurrently) - but I don't trust ourselves to always get that correct.